### PR TITLE
feat(ids-api): Add alive check to legal representative delegation type

### PR DESCRIPTION
**Pull request by valurefugl**
https://www.notion.so/Add-deceased-check-using-National-Registry-1115a76701d68092bb72fec8a14a94c4?pvs=4

## What

Refactor alive status check into separate service. Add check to legal representative delegation type.

Enable using national registry v3 api for the check.

## Why

So we don't show delegations for deceased individuals.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
  - Introduced the `AliveStatusService` for determining the alive or deceased status of individuals and companies based on national IDs.

- **Improvements**
  - Streamlined delegation services by integrating alive status checks directly, enhancing efficiency and reducing complexity.

- **Bug Fixes**
  - Improved error handling during status checks to ensure robust functionality without complete failures.

- **Documentation**
  - Updated module to include the new `AliveStatusService`, making it available for dependency injection.

### CODE REVIEW

File: libs/auth-api-lib/src/lib/delegations/alive-status.service.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          (identities[index] as IndividualDto)?.nationalId === nationalId,
+      )
+
+      return {
+        aliveNationalIds,
+        deceasedNationalIds,
+        aliveNameInfo: identitiesValuesNoError,
+      }
+    } catch (error) {
+      this.logger.error(`Error getting live status.}`, error)
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/alive-status.service.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 2]
```diff
@@ -335,11 +333,38 @@
...
-    const merged = records.map((d) =>
-      DelegationDTOMapper.recordToMergedDelegationDTO(d),
-    )
 
-    await Promise.all(merged.map((d) => this.updateName(d)))
+    const { aliveNationalIds, deceasedNationalIds, aliveNameInfo } =
+      await this.aliveStatusService.getStatus(
+        uniq(records.map((d) => d.fromNationalId)),
+      )
+
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts:341]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 3]
```diff
@@ -335,11 +333,38 @@
...
+            fromNationalId: delegation.fromNationalId,
+            toNationalId: delegation.toNationalId,
+            type: delegation.type,
+            provider: AuthDelegationProvider.DistrictCommissionersRegistry,
+          },
+          user,
+        ),
+      )
+
+      await Promise.all(deletePromises)
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 4]
```diff
@@ -335,11 +333,42 @@ export class DelegationsIncomingService {
...
+            },
+            user,
+          )
+        } catch (error) {
+          logger.error('Failed to remove delegation record.', error)
+        }
+      })
+
+      await Promise.all(deletePromises)
+    }
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts]: Refactored.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 5]
```diff
@@ -1,11 +1,8 @@
 import { BadRequestException, Injectable } from '@nestjs/common'
 import { InjectModel } from '@nestjs/sequelize'
+import uniq from 'lodash/uniq'
```
[the remark this answers is not in the document]
Comment (saevarma) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts:3]: ~ 👆

File: libs/auth-api-lib/src/lib/delegations/alive-status.service.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+import { UNKNOWN_NAME } from './constants/names'
+import { partitionWithIndex } from './utils/partitionWithIndex'
+
+export type NameInfo = {
+  nationalId: string
+  name: string
+}
+
+export class AliveStatusService {
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/alive-status.service.ts:22]: Pretty sure this was working. But added the decorator.

File: libs/auth-api-lib/src/lib/delegations/alive-status.service.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    private companyRegistryClient: CompanyRegistryClientService,
+    @Inject(LOGGER_PROVIDER)
+    private logger: Logger,
+  ) {}
+
+  /**
+   * Divides nationalIds into alive and deceased
+   * - Makes calls for every nationalId to NationalRegistry to check if the person exists.
+   * - Divides the nationalIds into alive and deceased, based on
+   *   1. All companies will be divided into alive.
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/alive-status.service.ts:35]: Agreed.

**Review by programm-ingovals — APPROVED** [review]
Regarding Personal Representative I see no issues.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts [hunk 8]
```diff
@@ -389,100 +395,11 @@ export class DelegationsIncomingCustomService {
...
-  }
-
   /**
    * Finds person by nationalId.
    */
   private getPersonByNationalId(
-    identities: Array<FromNameInfo | null>,
+    identities: Array<NameInfo | null>,
     nationalId: string,
   ) {
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming-custom.service.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts [hunk 9]
```diff
@@ -335,11 +333,42 @@ export class DelegationsIncomingService {
...
+
+      await Promise.all(deletePromises)
+    }
+
+    const merged = records
+      .filter((d) => aliveNationalIds.includes(d.fromNationalId))
+      .map((d) => ({
+        ...DelegationDTOMapper.recordToMergedDelegationDTO(d),
+        fromName: this.getNameFromNameInfo(d.fromNationalId, aliveNameInfo),
+      }))
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/delegations-incoming.service.ts]: Fixed.

File: libs/auth-api-lib/src/lib/delegations/alive-status.service.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+          // Pass through although Þjóðskrá API throws an error
+          identities[index] instanceof Error ||
+          // Make sure we can match the person to the nationalId, i.e. not deceased
+          (identities[index] as IndividualDto)?.nationalId === nationalId,
+      )
+
+      return {
+        aliveNationalIds,
+        deceasedNationalIds,
+        aliveNameInfo: identitiesValuesNoError,
```
[the remark this answers is not in the document]
Comment (valurefugl) [reply libs/auth-api-lib/src/lib/delegations/alive-status.service.ts:76]: Not sure I follow.. does this still apply with the latest changes?

**Review by saevarma — APPROVED** [review]
🙌

**Review by peturgq — APPROVED** [review]

**Pull request merged**